### PR TITLE
resolve bookmark reported by template

### PR DESCRIPTION
**Pull request by superyyrrzz**
@chenkennt @vwxyzh @ansyral @hellosnow @qinezh

### CODE REVIEW

File: src/docfx.website.themes/common/ManagedReference.common.js [hunk 1]
```diff
@@ -51,6 +51,20 @@ exports.transform = function (model)  {
   return model;
 }
 
+exports.getBookmarks = function (model)  {
+  if (!model) return null;
+
+  var bookmarks = {};
```
Comment (ansyral) [inline src/docfx.website.themes/common/ManagedReference.common.js]: >bookmarks [](start = 6, length = 9)

return?

File: src/Microsoft.DocAsCode.Utility/UriUtility.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        public static bool HasFragment(string uriString)
+        {
+            if (string.IsNullOrEmpty(uriString)) throw new ArgumentNullException(nameof(uriString));
+            return uriString.IndexOf(FragmentMarker) != -1;
+        }
+
+        public static string GetFragment(string uriString)
+        {
+            if (string.IsNullOrEmpty(uriString)) throw new ArgumentNullException(nameof(uriString));
+            var index = uriString.IndexOf(FragmentMarker);
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Utility/UriUtility.cs]: >FragmentMarker [](start = 42, length = 14)

no need to consider queryMarker?

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs [hunk 3]
```diff
@@ -580,8 +580,12 @@ private void FeedXRefMap(List<ManifestItemWithContext> manifest, IDocumentBuildC
                 using (new LoggerFileScope(m.FileModel.LocalPathFromRepoRoot))
                 {
                     Logger.LogDiagnostic($"Feed xref map from template for {m.Item.DocumentType}...");
-                    var bookmarks = m.Options.Bookmarks;
-                    // TODO: Add bookmarks to xref
+                    // TODO: use m.Options.Bookmarks directly after all templates report bookmarks
+                    var bookmarks = m.Options.Bookmarks ?? m.FileModel.Bookmarks;
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: >m.Options.Bookmarks [](start = 36, length = 19)

Question: the method FeedXRefMap seems to run before templateProcessor.Process? then how could templates collect bookmark info?

File: src/Microsoft.DocAsCode.Utility/UriUtility.cs (hunk 2 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Utility/UriUtility.cs]: Xref still not support query string for now. I'm consider add some other methods that also consider queryMarker later.

---
In reply to: [86513116](https://github.com/dotnet/docfx/pull/860#discussion_r86513116) [](ancestors = 86513116)

File: src/Microsoft.DocAsCode.Utility/UriUtility.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        public static string GetNonFragment(string uriString)
+        {
+            if (string.IsNullOrEmpty(uriString)) throw new ArgumentNullException(nameof(uriString));
+            var index = uriString.IndexOf(FragmentMarker);
+            return index == -1 ? uriString : uriString.Remove(index);
+        }
+
+    }
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Utility/UriUtility.cs]: remove extra spaces before this line?

File: src/Microsoft.DocAsCode.Utility/UriUtility.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            if (string.IsNullOrEmpty(uriString)) throw new ArgumentNullException(nameof(uriString));
+            return uriString.IndexOf(FragmentMarker) != -1;
+        }
+
+        public static string GetFragment(string uriString)
+        {
+            if (string.IsNullOrEmpty(uriString)) throw new ArgumentNullException(nameof(uriString));
+            var index = uriString.IndexOf(FragmentMarker);
+            return index == -1 ? string.Empty : uriString.Substring(index);
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Utility/UriUtility.cs]: >index == -1 [](start = 19, length = 11)

can call HasFragment?

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs (hunk 3 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: Template needn't collect, but decide bookmark info, then report it to DocFX for solving xref, and fill into mustache.

---
In reply to: [86513579](https://github.com/dotnet/docfx/pull/860#discussion_r86513579) [](ancestors = 86513579)

**Comment by qinezh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/XrefDetails.cs [hunk 6]
```diff
@@ -101,17 +101,16 @@ public void ApplyXrefSpec(XRefSpec spec)
...
             {
-                if (string.IsNullOrEmpty(Anchor))
-                {
-                    // TODO: hashtag from tempalte
-                    // TODO: What if href is not html?
-                    Anchor = "#" + GetHtmlId(Uid);
-                }
+                Anchor = UriUtility.GetFragment(spec.Href);
+            }
+            if (!string.IsNullOrEmpty(spec.Href))
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/XrefDetails.cs]: >!string.IsNullOrEmpty(spec.Href)) [](start = 16, length = 33)

merge the if with previous !string.IsNullOrEmpty(spec.Href)?

**Comment by hellosnow** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs (hunk 3 above)
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: The `FeedOptions` above calls `exports.getOptions` in js.

---
In reply to: [86703507](https://github.com/dotnet/docfx/pull/860#discussion_r86703507) [](ancestors = 86703507,86513579)

**Comment by ansyral** [comment]
:shipit:

**Pull request merged**